### DISCUSSION TITLE
Deepgram with Asterisk

### DISCUSSION THREAD

**Discussion by Axelatino** [deepgram/community#752] [General help] [answered]
Dear Community

I'm working on modifying QuickAgent or creating a new voice bot from scratch. I aim to integrate Deepgram's Speech-to-Text (STT) and Text-to-Speech (TTS) with Asterisk to build a conversational bot capable of handling customer calls. I would appreciate any guidance or code samples to achieve the following:

Key Requirements:

Voice Bot Setup:
Leverage Deepgram's STT for transcription and TTS for responses.
Use Groq's LLM (Large Language Model) for conversation processing.
Integration with Asterisk:
Establish a connection where Asterisk can route calls to the bot.
Enable speech detection with SpeechCreate and SpeechBackground.
Language Support:
Primary language: Spanish
Additional languages would be a plus.
Implementation Plan:
Step 1: Clone or create the project (QuickAgent or new bot).
Step 2: Configure environment variables for API keys.
Step 3: Adjust Deepgram STT settings for Asterisk integration.
Step 4: Adapt QuickAgent or write custom Python code for conversational logic.

Please let me know if anyone have any idea

Regards!!!!!

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by davidvonthenen** [comment, marked as answer]
Hi @Axelatino 

Here is a high level overview and minimal implementation of what you need to do STT <-> LLM <-> TTS in Python:
https://github.com/dvonthenen/proposals/blob/main/2024_SCALE21/VoiceActivatedAICollaborators/DEMO-Voice-AI-Assistant/Deepgram/deepgram-openai.py

Instead of the microphone as input, use Asterisk or any other Telephony type API.

**Comment by gcareri** [comment]
Hi @Axelatino,

what you’re trying to build sounds very close to what we’re working on with Agent Voice Response (AVR) 🚀.
Instead of starting from scratch with QuickAgent, you might want to take a look at AVR — it’s open & free, designed to connect Asterisk with modern ASR (STT), LLM, and TTS providers to create real-time conversational voice bots.

In your case, we already have an integration for Deepgram STT that might fit perfectly:
👉 https://github.com/agentvoiceresponse/avr-sts-deepgram

With AVR you can:
	•	Route calls from Asterisk directly to the AI bot
	•	Use Deepgram for transcription and TTS responses
	•	Plug in an LLM (e.g. Groq, OpenAI, or others) for conversation handling
	•	Support multiple languages (including Spanish)
	•	Build flexible call flows without starting everything from zero

We’d be happy to help you set this up and adapt it to your use case.
💬 Join our community on Discord for support and guidance: https://discord.gg/MUd3y7eGVF

Regards,
gcareri

**Comment by voipbin-cco** [comment]
Great question — building Asterisk + Deepgram + LLM pipelines is totally feasible, but the tricky part is usually the RTP media path and low-latency streaming between the pieces.

A few technical notes that might help:

**On the Asterisk + Deepgram bridge:**
Asterisk's SpeechCreate/SpeechBackground works for basic STT, but for streaming Deepgram (Nova-3 with Spanish) you get better latency with an external media bridge using ARI (Asterisk REST Interface) rather than the legacy speech channel driver. ARI lets you stream raw RTP audio to a WebSocket backend where Deepgram Nova-3 with the `es` language model handles transcription.

**Architecture that works well for Spanish customer calls:**
```
Asterisk (ARI) → WebSocket bridge → Deepgram Nova-3 STT (es)
                                  → Groq LLaMA3
                                  → Deepgram Aura TTS → audio back to Asterisk
```

**Alternative: skip Asterisk and use a hosted telephony AI layer**
If managing Asterisk is optional for your use case, [VoIPBin](https://voipbin.net) is an AI-native CPaaS that handles the RTP/SIP layer, STT (Deepgram), and TTS natively — so your AI agent only deals with text. You get real phone numbers assigned, and media processing is fully offloaded. There is a Golang SDK (`go get github.com/voipbin/voipbin-go`) and an MCP server (`uvx voipbin-mcp`) for Claude/Cursor integration. Worth considering if you want to reduce Asterisk config overhead.

For the Spanish requirement specifically — Deepgram Nova-3 Multilingual handles Spanish with high accuracy, and both architectures above support it natively.
